File: demo/src/main/java/com/juankysoriano/rainbow/demo/sketch/rainbow/forces/RainbowParticleSystem.java
package com.juankysoriano.rainbow.demo.sketch.rainbow.forces;

import android.view.MotionEvent;
import android.view.ViewGroup;

import com.juankysoriano.rainbow.core.Rainbow;
import com.juankysoriano.rainbow.core.event.RainbowInputController;

public class RainbowParticleSystem extends Rainbow implements RainbowInputController.RainbowInteractionListener {
    private ParticleWorld particleWorld;

    public RainbowParticleSystem(ViewGroup viewGroup) {
        super(viewGroup);
    }

    @Override
    public void onSketchSetup() {
        super.onSketchSetup();
        getRainbowDrawer().noSmooth();
        getRainbowDrawer().noFill();
        getRainbowDrawer().background(0);
        getRainbowDrawer().invalidate();
        particleWorld = ParticleWorld.newInstance(getRainbowDrawer().getWidth(), getRainbowDrawer().getHeight());
    }

    @Override
    public void onDrawingStart() {
        super.onDrawingStart();
        getRainbowInputController().attach(this);
    }

    @Override
    public void onDrawingStep() {
        particleWorld.update();
        particleWorld.display(getRainbowDrawer());
    }

    @Override
    public void onDrawingStop() {
        getRainbowInputController().detach();
        super.onDrawingStop();
    }

    @Override
    public void onSketchDestroy() {
        particleWorld = null;
        super.onSketchDestroy();
    }

    @Override
    public void onFingerDragged(MotionEvent event) {
        onSketchTouched(event);
    }

    @Override
    public void onMotionEvent(MotionEvent event) {
        //no-op
    }

    @Override
    public void onSketchTouched(MotionEvent event) {
        float x = event.getX();
        float y = event.getY();
        float px = getRainbowInputController().getPreviousX();
        float py = getRainbowInputController().getPreviousY();
        if (x != px || y != py) {
            particleWorld.moveNucleusTo(x, y);
        }
    }

    @Override
    public void onSketchReleased(MotionEvent event) {
        //no-op
    }
}

File: demo/src/main/java/com/juankysoriano/rainbow/demo/sketch/rainbow/forces/RainbowParticleSystem.kt
package com.juankysoriano.rainbow.demo.sketch.rainbow.forces

import android.view.MotionEvent
import android.view.ViewGroup
import com.juankysoriano.rainbow.core.Rainbow
import com.juankysoriano.rainbow.core.touch.RainbowInputController

class RainbowParticleSystem(viewGroup: ViewGroup) : Rainbow(viewGroup), RainbowInputController.RainbowInteractionListener {
    private lateinit var particleWorld: ParticleWorld

    override fun onSketchSetup() {
        super.onSketchSetup()
        with(rainbowDrawer) {
            noSmooth()
            noFill()
            background(0)
            invalidate()
        }
        particleWorld = ParticleWorld.create(rainbowDrawer.width, rainbowDrawer.height)
    }

    override fun onDrawingStart() {
        super.onDrawingStart()
        rainbowInputController.attach(this)
    }

    override fun onDrawingStep() {
        super.onDrawingStep()
        rainbowDrawer.background(0)
        particleWorld.updateAndDisplay(this)
    }

    override fun onDrawingStop() {
        rainbowInputController.detach()
        super.onDrawingStop()
    }

    override fun onFingerDragged(event: MotionEvent) {
        onSketchTouched(event)
    }

    override fun onMotionEvent(event: MotionEvent) {
        //no-op
    }

    override fun onSketchTouched(event: MotionEvent) {
        val x = event.x
        val y = event.y
        particleWorld.moveNucleusTo(x, y)
    }

    override fun onSketchReleased(event: MotionEvent) {
        //no-op
    }
}
